### DISCUSSION TITLE
Quick install of the btop binary

### DISCUSSION THREAD

**Discussion by tliron** [aristocratos/btop#173] [Ideas]
Here's an example for installing the latest version on Linux x86_64 (run as root):
```
VERSION=$(curl --silent https://api.github.com/repos/aristocratos/btop/releases/latest | grep '"tag_name":' | sed -E 's/.*"([^"]+)".*/\1/')
URL=https://github.com/aristocratos/btop/releases/download/$VERSION/btop-${VERSION:1}-x86_64-linux-musl.tbz
wget "$URL" -O - | tar --extract --bzip2 --directory=/usr/bin --strip=1 bin/btop
```
Maybe it's worth putting something like this in the README?

**Comment by userdocs** [comment]
If the release did not have version information in the asset you could just do this and it would always get the latest release for that asset.

https://github.com/aristocratos/btop/releases/latest/download/btop-aarch64-linux-musl.tbz

Currently it is like this - a working example but not latest release friendly since the version 1.1.2 is hardcoded to the URL.

https://github.com/aristocratos/btop/releases/latest/download/btop-1.1.2-aarch64-linux-musl.tbz

I think this is the best and most modern way to do this

Using the api is ok but it has rate limits for unauthorised calls of 60 per IP hour. So not 100% Reliable in some settings.

Since you are setting a VERSION var then reusing the var i suggest you use `git ls-remote` and `awk` to refine the search since this is not limited by api quota.

```
git ls-remote -q -t --refs https://github.com/aristocratos/btop.git | awk '/v/{sub("refs/tags/", ""); print $2 }' | awk '!/^$/' | sort -rV | head -n 1
```

Giving 

```
v1.1.2
```

**Reply by tliron** [reply]
Thanks, the ls-remote is a neat trick!

**Comment by aristocratos** [comment]
@tliron 
I would not recommend installing this way since you're ignoring the included Makefile installer and only extracting the binary, consequently not getting any updates to the themes or documentation (or future included files). And if installing first time you would not have any themes or documentation at all.

**Comment by tliron** [comment]
Well then, a better installation script is in order until RPM/DEB/etc. packaging is available.

**Reply by aristocratos** [reply]
`./install.sh` or `make install` isn't easy enough?

**Reply by tliron** [reply]
Having to compile everything locally, with the requirement of having a full compilation stack and dependencies, seems like a lot to me. But it's your choice as maintainer as to how widely you want your project to be available.

**Reply by aristocratos** [reply]
Read https://github.com/aristocratos/btop#installation
The release packages include binaries, themes, documentation and a installer in form of a Makefile that installs all files following system wide or manually set PREFIX path. It also includes an easy way to set SUID bit for the binary.

The only time you would need to compile is if you don't have binaries available for your platform.

**Reply by tliron** [reply]
I see, so really only Make is required as a dependency. I still think this is too hard for most users, but I guess they are not your target audience anyway.

In any case, a one-liner to download and run `./install.sh` could make things easier.

**Reply by aristocratos** [reply]
I really don't see how the current instructions are too hard, you can literally do them without even opening a terminal.
1. Download the appropriate .tbz file for your platform and architecture.
2. Unpack to a new folder (if instructions for how to unpack compressed files is needed, then yes, that would not be the target audience for this terminal program).
3. Double click (if so inclined) the `install.sh` file
4. Done

Sure, there is the possibility that `Make` is not installed, but I'm pretty sure the "big" distros actually give instructions in the terminal for how to install `Make` if it's missing.

Wouldn't be an issue to rewrite the installer to a simple shell script but no one has yet complained about not being able to run the installer.

**Reply by tliron** [reply]
Thank you for explaining your position and for making this utility.

**Reply by aristocratos** [reply]
Thank you for sharing your ideas, I'm sure the code snippet you shared will come to use for people who for example want to set up their own btop update script.

There will eventually be more distro packages and from indications I've gotten, for debian and ubuntu at least. But wanna make sure btop is without a doubt stable before I encourage distribution packagers for big distros that won't be that fast in updating the releases.

**Reply by userdocs** [reply]
@aristocratos I do think that providing release assets without version information in them will provide a consistent and easy method to get the latest archive that uses a github built in feature. This is how I do it and I will always get the latest release.

```
wget https://github.com/userdocs/btop-crossbuilds/releases/latest/download/x86_64-linux-musl.tar.gz
```

By having version Info hardcoded into the archive name it is not taking advantage if the latest release URL redirection Github does.

So I still think this part of my comment is a valid idea that works for both parties.

**Reply by userdocs** [reply]
And just to provide an example in action, this is what it looked like for me to update my installation using my repo.

note: in the example my archive structure is different to yours so the unpack command is relevant to demonstrate the example but it's not required that you also change your archive structure as part of the URL format change.

```
btop -v
btop version: 1.1.1
```

```
wget -q "https://github.com/userdocs/btop-crossbuilds/releases/latest/download/x86_64-linux-musl.tar.gz" -O ~/btop.tar.gz
tar xf "btop.tar.gz" --strip-components=1 -C ~/
```

```
btop -v
btop version: 1.1.2
```

**Reply by aristocratos** [reply]
@userdocs 
Sure, I see no issue in removing the version number from the release files. Will make the changes to my release script and next release will be in the format of `btop-x86_64-linux-musl.tbz`.

**Reply by userdocs** [reply]
I'll try this again, 

We can now do this

Extract relative to where we are in the shell

```bash
wget -q -O- "https://github.com/aristocratos/btop/releases/latest/download/btop-x86_64-linux-musl.tbz" | tar -xjf -
```

Extract to a pre existing path.

```bash
wget -q -O- "https://github.com/aristocratos/btop/releases/latest/download/btop-x86_64-linux-musl.tbz" | tar -xjf - -C ~/some_dir
```

Extract a specific file and path from the archive and ignore the rest.

```bash
wget -q -O- "https://github.com/aristocratos/btop/releases/latest/download/btop-x86_64-linux-musl.tbz" | tar -xjf - -C ~/some_dir bin/btop
```

**Comment by alexdelprete** [comment]
Sorry to bump this, I'm trying to integrate btop installation in my dotfiles config.

If I use
```
https://github.com/aristocratos/btop/releases/latest/download/btop-x86_64-linux-musl.tbz
```
or
```
https://github.com/userdocs/btop-crossbuilds/releases/latest/download/x86_64-linux-musl.tar.gz
```

I always get `v1.2.7`, while latest is v1.2.9. Is there an URL that can provide the latest version?

Thanks for creating btop, @aristocratos. :)

**Reply by userdocs** [reply]
How are you determining the version?

If I do this

```
wget -q -O- "https://github.com/aristocratos/btop/releases/latest/download/btop-x86_64-linux-musl.tbz" | tar -xjf - -C ~/
~/bin/btop --version
```

I will see:

```
btop version: 1.2.9
```

I guess if version is giving you the wrong result, you have 1.2.7 somewhere else in your path.

**Reply by aristocratos** [reply]
`https://github.com/aristocratos/btop/releases/latest/download/btop-x86_64-linux-musl.tbz` gets me the latest version `v1.2.9`.

Sounds like an caching issue on your end. You got any custom caching settings setup for your system or network?

**Comment by alexdelprete** [comment]
Sorry @userdocs & @aristocratos, found the problem: on the system I was testing with, btop was installed some months ago through apt, and it's 1.2.7. :)

The url `https://github.com/aristocratos/btop/releases/latest/download/btop-x86_64-linux-musl.tbz` is correct.

One question: for a full installation, the advice is to extract the archive and execute the install script, right? extracting only the binary I would lose some functionality if I understood correctly.

Thanks for the help.

**Reply by aristocratos** [reply]
Running the install script or `make install` would be recommended otherwise you wouldn't get potential theme updates. If you had previously set the SUID bit for btop (always starting with admin rights) you will also need to again run `make setuid`.

**Comment by alexdelprete** [comment]
Done, works beautifully and silently. Thanks a lot for the help. Love the interface, you made an amazing job.

```bash
echo "Installing/Upgrading btop++"
mkdir ./btopinst &> /dev/null
pushd ./btopinst &> /dev/null
wget -q https://github.com/aristocratos/btop/releases/latest/download/btop-x86_64-linux-musl.tbz -O btop.tbz
tar -xjvf btop.tbz &>/dev/null
make -s install &>/dev/null
popd  &> /dev/null
rm -rf ./btopinst
```

**Comment by MatteoGauthier** [comment]
Another version of an installation script

```bash
mkdir -p btop-install/output
wget -qO - https://github.com/aristocratos/btop/releases/latest/download/btop-x86_64-linux-musl.tbz | tar -xj -C btop-install/output
(cd btop-install/output/btop && sudo make install && sudo make setuid)
rm -rf btop-install
```
